### PR TITLE
Mark deleted before drop index

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #17408 #17521 
Signed-off-by: Cai.Zhang <cai.zhang@zilliz.com>

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/rootcoord/meta_table.go [hunk 1]
```diff
@@ -973,6 +1037,138 @@ func (mt *MetaTable) GetInitBuildIDs(collName, indexName string) ([]UniqueID, er
...
+
+	mt.ddLock.Unlock()
+	if err := mt.RemoveSegments(collID, partID, recycledSegIDs); err != nil {
+		log.Warn("remove segments failed, wait to retry", zap.Int64("collID", collID), zap.Int64("partID", partID),
+			zap.Int64s("segIDs", recycledSegIDs), zap.Error(err))
+	}
+}
+
+func (mt *MetaTable) RemoveSegments(collID, partID UniqueID, segIDs []UniqueID) error {
+	mt.ddLock.Lock()
```
Comment (xiaofan-luan) [inline internal/rootcoord/meta_table.go:1067]: add a log about which segment to remove

File: internal/rootcoord/meta_table.go [hunk 2]
```diff
@@ -973,6 +1037,138 @@ func (mt *MetaTable) GetInitBuildIDs(collName, indexName string) ([]UniqueID, er
...
+	recycledSegIDs := make([]UniqueID, 0)
+	if segMap, ok := mt.partID2SegID[partID]; ok {
+		for segID := range segMap {
+			if _, ok := segIDs[segID]; !ok {
+				recycledSegIDs = append(recycledSegIDs, segID)
+			}
+		}
+	}
+
+	mt.ddLock.Unlock()
```
Comment (xiaofan-luan) [inline internal/rootcoord/meta_table.go:1059]: the lock handling is weird.
Usually we can RemoveSegments and RemoveSegmentsInternal where RemoveSegmentsInternal don't get lock

File: internal/rootcoord/meta_table.go [hunk 3]
```diff
@@ -973,6 +1037,138 @@ func (mt *MetaTable) GetInitBuildIDs(collName, indexName string) ([]UniqueID, er
...
+		v, err := proto.Marshal(clonedCollMeta)
+		if err != nil {
+			log.Error("MetaTable RecycleDroppedIndex Marshal collMeta failed",
+				zap.String("key", k), zap.Error(err))
+			return err
+		}
+		saveMeta[k] = string(v)
+
+		// TODO: If there are 64 or more indexes on a collection, the number of ops here will be greater than 128, and etcd will thrown an error
+		if err = mt.txn.MultiSaveAndRemoveWithPrefix(saveMeta, delMeta); err != nil {
```
Comment (xiaofan-luan) [inline internal/rootcoord/meta_table.go:1146]: Need to handle cleanup in batch, otherwise it will never cleanup

**Comment by xiaofan-luan** [comment]
Overall lgtm, a few issues to handle

**Comment by xiaocai2333** [comment]
issue: #17521

File: internal/rootcoord/meta_table.go [hunk 4]
```diff
@@ -794,7 +793,7 @@ func (mt *MetaTable) DeletePartition(collID typeutil.UniqueID, partitionName str
 	return partID, nil
 }
 
-// AddIndex add index
+// AddIndex add index; before calling this function, you must check whether indexing is required
 func (mt *MetaTable) AddIndex(segIdxInfo *pb.SegmentIndexInfo) error {
```
Comment (xiaofan-luan) [inline internal/rootcoord/meta_table.go:797]: what happend if addIndex but the indexID has been dropped?

File: internal/rootcoord/root_coord.go [hunk 5]
```diff
@@ -369,84 +369,129 @@ func (c *Core) checkFlushedSegmentsLoop() {
...
+			}
+		}
+	}
+}
+
+func (c *Core) createIndexForSegment(ctx context.Context, collID, partID, segID UniqueID, numRows int64, binlogs []*datapb.FieldBinlog) error {
+	collID2Meta, _, indexID2Meta := c.MetaTable.dupMeta()
+	collMeta, ok := collID2Meta[collID]
+	if !ok {
+		log.Error("collection meta is not exist", zap.Int64("collID", collID))
```
Comment (xiaofan-luan) [inline internal/rootcoord/root_coord.go:404]: shoud move the check into AddIndex

**Comment by xiaocai2333** [comment]
issue: #17611

**Pull request merged**